### ISSUE TITLE
New features and options in mixed version clusters

### ISSUE THREAD

**Issue by javanna** [elastic/elasticsearch#70805]
We've been discussing as part of #70784 how we can handle scenarios where a certain search feature is added in a certain version, and CCS is used to search across clusters that don't necessarily support such feature.

A similar problem manifests in mixed version clusters, where any feature or option is added that not all nodes support. Up until now we handled most of this at the transport layer, by not serializing the bits that older nodes cannot deserialize, which leads to a lenient/degraded behaviour. This is also evaluated case by case depending on the feature/option that is being added, but I think that it would make sense to discuss best practices and what we want to guarantee in a mixed version cluster.

While mixed version clusters are a temporary situation and degraded functionality may be acceptable, there are cases where returning errors or at least emit some kind of warning could make sense.

We are stumbling upon this in problem in #69948, where we are adding a new dynamic template format, as well as a new option to the bulk request. We are discussing failing indexing requests that include the new option if such option cannot be honored, depending on the node that ends up executing such shard request. Also, we are contemplating accepting the new dynamic template format only once all nodes are on the version that supports it. I would like to validate these proposals and have a wider discussion around how to approach this kind of problem.

**Comment by DaveCTurner** [comment]
> Up until now we handled most of this at the transport layer, by not serializing the bits that older nodes cannot deserialize, which leads to a lenient/degraded behaviour.

I don't think that's universally true, in many cases if we encounter a request that cannot be serialised then we aren't lenient, we simply fail. Here's the first three examples of that I found:

https://github.com/elastic/elasticsearch/blob/3a64ecb6bfe5c7d88ac73954a3b498cc71ab8d6e/server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/delete/DeleteSnapshotRequest.java#L84-L92

https://github.com/elastic/elasticsearch/blob/28710c985d7f294a38bcf666a367449e5c486578/server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/restore/RestoreSnapshotRequest.java#L126-L131

https://github.com/elastic/elasticsearch/blob/5fcda57b37f1d87de116c98b8234affbc29500d1/server/src/main/java/org/elasticsearch/action/admin/indices/shrink/ResizeRequest.java#L112-L117

Failing a request that uses new features in an incompletely-upgraded cluster is completely reasonable IMO. There are definitely nicer ways to degrade in some cases, but we shouldn't simply drop parts of requests to fit them into an older wire format if it changes the meaning of the request.

> depending on the node that ends up executing such shard request

It's really tricky to make this a function of the _node_ version, because nodes come and go in a manner that races with indexing. Can we use the index created version instead? We guarantee that an index created in version N is always allocated to nodes with version >N.

Julie sent an email out to the internal es@ mailing list 2021-01-07 which clarifies a similar point about using the index-created version for BWC decisions if possible.

> Also, we are contemplating accepting the new dynamic template format only once all nodes are on the version that supports it. 

This would be fragile, we do not guarantee that `DiscoveryNodes#getMinNodeVersion()` only ever increases. If you're in the middle of a rolling upgrade and the last old-version node temporarily leaves the cluster due to a network wobble then you'll be in a state where all nodes have been upgraded, but then the last old-version node rejoins and you're back to having a non-upgraded node. We do protect against going backwards across a major version upgrade, but not otherwise. We could change things to be stricter there, this isn't the first time this has come up, and it is a bit trappy for us developers.

**Comment by javanna** [comment]
I was referring to some old habit of skipping whatever can not be serialized, and I am happy to hear that we have already started failing when it makes sense. That is clearer to users in my opinion compared to ignoring part of their requests.

I had a chat with @DaveCTurner on another channel and we concluded that: failing at transport is an option but there are other possibly more graceful ways to fail, for instance using the index created version, and relying on the guarantee that a shard of an index created on version X will never be allocated on a node on version < X.

For the specific case of #69948 then, we should accept a bulk request holding the new option only for indices that are created on (or after) the version that supports the new option, and this guarantees early failures (on the coordinating node) and consistent behaviour. The same can be done around accepting the enhanced dynamic template: accept it only for indices created on (or after) the version that supports the new feature first.

**Comment by jpountz** [comment]
> For the specific case of #69948 then, we should accept a bulk request holding the new option only for indices that are created on (or after) the version that supports the new option

I'm not sure if this is practical. For instance, if you are indexing into a data stream, the data shipper has no idea of the index-created version of the current backing index since it depends on the rollover schedule. For this case, I'd rather fail on the transport and require data shippers to wait until the cluster is fully upgraded before starting to use new features, which comes for free given our requirement to fully upgrade Elasticsearch before upgrading data shippers (e.g. Beats or Agent). Am I missing something?

**Comment by javanna** [comment]
I agree @jpountz what sounds not practical is that a new index needs to be created once all nodes are upgraded for the new feature to be usable, while it is perfectly fine to use it on older indices that are allocated to newer nodes.

**Comment by javanna** [comment]
I opened this issue a long while ago for discussion. Since then, we have consolidated our approach when it comes to making changes to the search API and other API that support cross-cluster calls: throw error when serializing to a node that does not support the feature, never silently fail. This cannot be applied to all API, in other scenarios we go case by case: we generally try to avoid inconsistencies and to be explicit with users. I would tend to close this issue, unless others think that there is work left to do around it.
